borg/platform_windows.pyx

```diff
@@ -0,0 +1,291 @@
+    void SetLastError(DWORD)
+
+    BOOL InitializeSecurityDescriptor(BYTE*, DWORD)
+
+    BOOL LookupAccountNameW(LPCTSTR, LPCTSTR, PSID, LPDWORD, LPCTSTR, LPDWORD, LPDWORD)
+    BOOL GetSecurityDescriptorDacl(PSID, BOOL*, _ACL**, BOOL*)
+
+    cdef extern int ERROR_INSUFFICIENT_BUFFER
+    cdef extern int ERROR_INVALID_SID
+
+    cdef extern int OWNER_SECURITY_INFORMATION
+    cdef extern int GROUP_SECURITY_INFORMATION
+    cdef extern int DACL_SECURITY_INFORMATION
+    cdef extern int SACL_SECURITY_INFORMATION
+    cdef extern int LABEL_SECURITY_INFORMATION
+    cdef extern int ATTRIBUTE_SECURITY_INFORMATION
+    cdef extern int SCOPE_SECURITY_INFORMATION
+    cdef extern int BACKUP_SECURITY_INFORMATION
+    cdef extern int UNPROTECTED_SACL_SECURITY_INFORMATION
+    cdef extern int UNPROTECTED_DACL_SECURITY_INFORMATION
+    cdef extern int PROTECTED_SACL_SECURITY_INFORMATION
+    cdef extern int PROTECTED_DACL_SECURITY_INFORMATION
+
+    cdef extern int SECURITY_DESCRIPTOR_MIN_LENGTH
+
+cdef extern from 'accctrl.h':
+    ctypedef enum _SE_OBJECT_TYPE:
+        SE_FILE_OBJECT
+    ctypedef _SE_OBJECT_TYPE SE_OBJECT_TYPE
+    struct _TRUSTEE_W:
+        uint16_t TrusteeForm
+        uint16_t TrusteeType
+        LPCTSTR ptstrName
+
+    struct _EXPLICIT_ACCESS_W:
+        DWORD grfAccessPermissions
+        uint16_t grfAccessMode
+        DWORD grfInheritance
+        _TRUSTEE_W Trustee
+
+    cdef extern uint16_t TRUSTEE_IS_SID
+    cdef extern uint16_t TRUSTEE_IS_NAME
+    cdef extern uint16_t TRUSTEE_BAD_FORM
+
+    DWORD GetExplicitEntriesFromAclW(_ACL*, uint32_t*, _EXPLICIT_ACCESS_W**)
+
+cdef extern from 'Sddl.h':
+    ctypedef int* LPBOOL
+
+    BOOL GetFileSecurityW(LPCTSTR, int, PSID, DWORD, LPDWORD)
+    BOOL GetSecurityDescriptorOwner(PSID, PSID*, LPBOOL)
+    BOOL LookupAccountSidW(LPCTSTR, PSID, LPCTSTR, LPDWORD, LPCTSTR, LPDWORD, uint16_t*)
+    BOOL ConvertSidToStringSidW(PSID, LPCTSTR*)
+    BOOL ConvertStringSidToSidW(LPCTSTR, PSID*)
+    BOOL ConvertSecurityDescriptorToStringSecurityDescriptorW(BYTE*, DWORD, int, LPCTSTR*, int*)
+
+    cdef extern int SDDL_REVISION_1
+
+cdef extern from 'Aclapi.h':
+    ctypedef void* PACL
+    DWORD GetNamedSecurityInfoW(LPCTSTR, SE_OBJECT_TYPE, DWORD, PSID*, PSID*, PACL*, PACL*, _ACL**)
+    DWORD SetNamedSecurityInfoW(LPCTSTR, int, int, PSID, PSID, PACL, PACL)
+    DWORD SetEntriesInAclW(unsigned int, _EXPLICIT_ACCESS_W*, PACL, _ACL**)
+
+cdef PSID _get_file_security(filename, int request):
+    cdef DWORD length = 0
+    # N.B. This query may fail with ERROR_INVALID_FUNCTION
+    # for some filesystems.
+    cdef wchar_t* wcharfilename = PyUnicode_AsWideCharString(filename, NULL)
+    if wcharfilename == NULL:
+        print 'filename is NULL'
+    GetFileSecurityW(wcharfilename, request, NULL, 0, &length)
+    if GetLastError() == ERROR_INSUFFICIENT_BUFFER:
+        SetLastError(0)
+    else:
+        print '_get_file_security failed. Windows error:', GetLastError()
+        return NULL
```

Comment:
To make error handling easier you could define a function like this:

```py
def raise_error(failed_api='', path=''):
    cdef wchar_t *error_message
    error = GetLastError()
    if not error:
        return
    FormatMessageW(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
                   0, error, 0, <wchar_t*> &error_message, 0, 0)
    error_string = PyUnicode_FromWideChar(error_message, -1)
    error_string = (failed_api + ': ' + error_string).strip()
    if path:
        raise OSError(error, error_string, path)
    else:
        raise OSError(error, error_string)
```

(untested)